keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,297 @@
+
+In the traditional RPC client-server tracing model, a trace context is attached to a single incoming request, and is propagated with all requests the server makes to other servers required to fulfill the initial single request.  Conceptually, this proposal suggests treating a kubernetes cluster as a single RPC server.  The difference is that we attach context to objects, and propagate this context to objects modified as a result of the initial object modification.  For example, if a user creates a ReplicaSet, the kube-controller-manager will create many Pod objects as a result, and will propagate the context used to create the ReplicaSet to Pod objects as well.  This ensures that all actions taken by kubernetes controllers as a result of the initial user action are linked by the same context.
+
+For the alpha phase, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`.  As noted in [Tracing API Requests](#tracing-api-requests) above, storing the trace context with the context is _in addition_ to attaching a context to http requests to the apiserver.  The reason for this is explained in the [Controller Behavior](#controller-behavior) section below.  In some scenarios, controllers will want to update the trace context from A -> B, but want to associated that Update request with context A.
+
+This means two trace contexts are sent in different forms with Create/Patch/Update requests to the apiserver.  A trace context is around 32 bytes (16 bytes for the trace ID, 8 bytes for the span ID, and some metadata). See the [w3c spec](https://w3c.github.io/trace-context/#tracestate-field) for details.
+
+
+This annotation value is removed when an object's trace ends, to achieve the desired behavior from [section one](#trace-lifecycle).  For core kubernetes components, this must be done in the same request to the API Server as the status update which updates the object to its desired state.  This is a requirement to ensure tracing does not affect the scalability of kubernetes.  For other components, it is recommended, but not required to update the trace annotation in the same request.
+
+This proposal chooses to use annotations to store the SpanContext associated with an object.  This mirrors how trace context propagation is done with golang context.Context and http headers, which are both key/value stores.
+
+#### Controller Behavior
+
+When reconciling an object `Foo` a Controller must:
+
+1. Send the trace context stored in `Foo` in the http request context for all API requests. See [Tracing API Requests](#tracing-api-requests)
+1. Store the trace context of `Foo` in object `Bar` when updating the Spec of `Bar`. See [Propagating Context Through Objects](#propagating-context-through-objects)
+1. Export a span around work that attempts to drive the actual state of an object towards its desired state
+1. Replace the trace context of `Foo` when updating `Foo`'s status to the desired state
```

Comment:
There's one caveat in the controllers that was not covered here. Since controllers are edge driven, if user(s) modify an object twice, one after another, this will generate two update events, both noticed by the controller code. But upon processing the controller will process the latest state of the object. For example:
1. user A sets replicas 2
2. user B sets replicas 3
3. controller processes update and reads replicas 3 (the last state)
4. controller processes 2nd event and notices it's already fulfilled and does nothing.
Based on the above your tracing event 1 will be passed down to newly created pods, but you might loose event 2.
It's not always but something that should be accounted for.

Reply:
See the paragraph that starts with "Components should plumb..." a few paragraphs down.  Controllers should use the context associated with the desired state to which they are updating.  In your example, Step 3 would actually use the context from the second update, since that is the state to which it is updating.

This does seem to be another case where linking the previous "unfinished" trace to the new one would be helpful.  That isn't currently part of the proposal, but is something that has come up as a potential solution to scenarios in which an update occurs before an object reaches its desired state.

Reply:
We've dealt with similar issue when modeling state machines. When using tracing APIs like OpenTracing or OpenTelemetry, it is possible to record links to other traces when starting a new span. So if the trace context is stored in an "object" and is later updated to another trace context, they can be linked at that point, capturing the information in the trace about all events, even when replicas=2 was not actually executed.